urllib3/util/connection.py

```diff
@@ -1,14 +1,54 @@
 import socket
 try:
-    from select import poll, POLLIN
+    from select import poll, POLLIN, POLLOUT, POLLERR
 except ImportError:  # `poll` doesn't exist on OSX and other platforms
     poll = False
     try:
-        from select import select
+        from select import select as select
     except ImportError:  # `select` doesn't exist on AppEngine.
         select = False
 
 
+def _select(readlist, writelist, exceptionallist, timeout):
+    # Some platforms, e.g., OSX do not support poll
+    if not poll:
+        if select:
+            return select(readlist, writelist, exceptionallist, timeout)
+
+    # Using poll is better on platforms that support it
+    p = poll()
+    for fd in readlist:
+        p.register(fd, POLLIN)
+
+    for fd in writelist:
+        p.register(fd, POLLOUT)
+
+    for fd in exceptionallist:
+        p.register(fd, POLLERR)
+
+    return p.poll(timeout)
+
+
+def wait_to_read_data(socket, timeout=0.0):
+    sockets = _select([socket], [], [], timeout)
+
+    # select.select returns ([...], [...], [...])
+    if isinstance(sockets, tuple):
+        return sockets[0]
+
+    return sockets
+
+
+def wait_to_write_data(socket, timeout=0.0):
+    sockets = _select([], [socket], [], timeout)
+
+    # select.select returns ([...], [...], [...])
+    if isinstance(sockets, tuple):
+        return sockets[1]
+
+    return sockets
```

Comment:
A list of tuples (from poll). Poll returns a list of things like

```
[(3, POLLIN), (6, POLLOUT), ...]
```

Where `POLLIN` and `POLLOUT` are the actual integers that `select.POLL(IN|OUT)` are assigned to in the module, and `3` and `6` are file descriptors.